lib/smart_name.rb

```diff
@@ -112,186 +125,22 @@ def == other
-  end
-
-  def right_name
-    @right_name ||= right && self.class.new(right)
-  end
-
-  # Note that all names have a trunk and tag,
-  # but only junctions have left and right
-
-  def trunk
-    @trunk ||= simple? ? s : left
-  end
-
-  def tag
-    @tag ||= simple? ? s : right
-  end
-
-  def trunk_name
-    @trunk_name ||= simple? ? self : left_name
-  end
-
-  def tag_name
-    @tag_name ||= simple? ? self : right_name
-  end
-
-  def part_names
-    @part_names ||= parts.map(&:to_name)
-  end
-
-  def piece_names
-    @piece_names ||= pieces.map(&:to_name)
-  end
-
-  def pieces
-    @pieces ||=
-      if simple?
-        [self]
-      else
-        junction_pieces = []
-        parts[1..-1].inject parts[0] do |left, right|
-          piece = [left, right] * self.class.joint
-          junction_pieces << piece
-          piece
-        end
-        parts + junction_pieces
-      end
-  end
-
-  # ~~~~~~~~~~~~~~~~~~~~ SHOW / ABSOLUTE ~~~~~~~~~~~~~~~~~~~~
-
-  def to_show *ignore
-    ignore.map!(&:to_name)
-
-    show_parts = parts.map do |part|
-      reject = (part.empty? || (part =~ /^_/) || ignore.member?(part.to_name))
-      reject ? nil : part
-    end
-
-    show_name = show_parts.compact.to_name.s
-
-    case
-    when show_parts.compact.empty? then  self
-    when show_parts[0].nil?        then  self.class.joint + show_name
-    else show_name
-    end
-  end
-
-  def to_absolute context, args={}
-    context = context.to_name
-    parts.map do |part|
-      new_part =
-        case part
-        when /^_user$/i
-          name_proc = self.class.session
-          name_proc ? name_proc.call : part
-        when /^_main$/i            then self.class.params[:main_name]
-        when /^(_self|_whole|_)$/i then context.s
-        when /^_left$/i            then context.trunk
-          # note - inconsistent use of left v. trunk
-        when /^_right$/i           then context.tag
-        when /^_(\d+)$/i
-          pos = $~[1].to_i
-          pos = context.length if pos > context.length
-          context.parts[pos - 1]
-        when /^_(L*)(R?)$/i
-          l_s, r_s = $~[1].size, !$~[2].empty?
-          l_part = context.nth_left l_s
-          r_s ? l_part.tag : l_part.s
-        # when /^_/
-        #   custom = args[:params] ? args[:params][part] : nil
-        #   custom ? CGI.escapeHTML(custom) : part
-        # why are we escaping HTML here?
-        else
-          part
-        end.to_s.strip
-      new_part.empty? ? context.to_s : new_part
-    end * self.class.joint
-  end
-
-  def to_absolute_name *args
-    self.class.new to_absolute(*args)
-  end
-
-  def nth_left n
-    # 1 = left; 2= left of left; 3 = left of left of left....
-    (n >= length ? parts[0] : parts[0..-n - 1]).to_name
+  # @return true if name has a chain of parts that equals `subname`
+  def include? subname
+    subkey = subname.to_name.key
+    joint = Regexp.quote self.class.joint
+    key =~ /(^|#{joint})#{Regexp.quote subkey}($|#{joint})/
```

Comment:
Shouldn't this be looking for key equality in a part? key.parts (split joint).find{|sk| sk == subkey}